File: 0155_Min_Stack/MinStack.java
import java.util.EmptyStackException;
import java.util.Stack;

public class MinStack {
    private Stack<Integer> stack;
    private int min = Integer.MAX_VALUE;
    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack<>();
    }

    public void push(int x) {
        if (x < min) {
            // also push the current min under new
            // min in order to update min after pop
            if (!stack.empty()){
                stack.push(min);
            }
            min = x;
        }
        stack.push(x);
    }

    public void pop() {
        if (min == stack.pop()) {
            min = stack.empty() ? Integer.MAX_VALUE : stack.pop();
        }
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        if (stack.empty()) {
            throw new EmptyStackException();
        }
        return min;
    }
}

File: 0155_Min_Stack/MinStack.py
import sys


class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = list()
        self.min = sys.maxint

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x <= self.min:
            # store current min in order to update
            # when minimal number is popped
            if not len(self.stack) == 0:
                self.stack.append(self.min)
            self.min = x

        self.stack.append(x)

    def pop(self):
        """
        :rtype: void
        """
        if not len(self.stack) == 0:
            if self.min == self.stack.pop():
                    self.min = self.stack.pop() if not len(self.stack) == 0 else sys.maxint

    def top(self):
        """
        :rtype: int
        """
        return self.stack[len(self.stack) - 1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min
